Automatic request retries

ob-stripe opened the pull request:
r? @brandur-stripe @remi-stripe 
cc @stripe/api-libraries 

Add support for automatic request retries.

The implementation is based on stripe-go, so the retry conditions are the same: requests are retried on connection errors (timeouts) and 409's. Given the recent discussions in e.g. https://github.com/stripe/stripe-node/pull/559, let me know if you want me to change the conditions (e.g. to retry on 500's).

ob-stripe commented on src/Stripe.net/Infrastructure/Public/SystemNetHttpClient.cs:
```diff
@@ -24,6 +27,10 @@ public class SystemNetHttpClient : IHttpClient
 
+        private readonly object randLock = new object();
+
+        private readonly Random rand = new Random();
```
This random number generator is used to generate the jitter in retry delays. Microsoft recommends creating a single generator instance rather than creating a new instance every time, but also warns that `Random` is not thread-safe, so a lock is needed. Cf. https://docs.microsoft.com/en-us/dotnet/api/system.random?view=netframework-4.7.2#Multiple.

brandur-stripe replied:
Nice research. (And wow, a thread-safe convenient wrapper for `Rand` seems like a pretty obvious improvement for a future version of .NET.)

ob-stripe commented on src/Stripe.net/Infrastructure/Public/SystemNetHttpClient.cs:
```diff
@@ -60,24 +67,62 @@ public static System.Net.Http.HttpClient BuildDefaultSystemNetHttpClient()
+            int retry = 0;
+            OperationCanceledException timeoutException = null;
+
+            this.requestTelemetry.MaybeAddTelemetryHeader(request.StripeHeaders);
```
I've changed `RequestTelemetry.MaybeAddTelemetryHeader()` to add the header to the `StripeRequest` instance rather than the `HttpRequestMessage` instance. This is because the `HttpRequestMessage` instance is consumed by every request attempt, so we need to generate a new one from the `StripeRequest` instance for each retry. This way, we don't lose metrics when a request fails.

ob-stripe commented on src/StripeTests/Functional/TelemetryTest.cs:
```diff
@@ -53,12 +53,12 @@ public void TelemetryEnabled()
             this.ResetStripeClient();
             var fakeServer = FakeServer.ForMockHandler(this.MockHttpClientFixture.MockHandler);
-            fakeServer.DelayMilliseconds = 10;
+            fakeServer.Delay = TimeSpan.FromMilliseconds(20);
```
This test was flaky and randomly failed. I think it's because the resolution of the timer used internally by `Task.Delay` is only around ~15 milliseconds, so even when the fake server was supposed to sleep for 10 milliseconds, we sometimes measured a duration of 8-9 milliseconds and the call to `Verify` would fail because we asserted that `d >= 10`.

I've relaxed the test by sleeping for a bit longer (20 milliseconds) and asserting that the duration is >= 15.

ob-stripe commented on src/Stripe.net/Infrastructure/Public/SystemNetHttpClient.cs:
```diff
@@ -60,24 +67,62 @@ public static System.Net.Http.HttpClient BuildDefaultSystemNetHttpClient()
+                        .ConfigureAwait(false);
+                }
+                catch (OperationCanceledException exception)
+                    when (!cancellationToken.IsCancellationRequested)
```
Internally, `HttpClient` uses a `CancelationToken` to handle timeouts, so we need to use a `when` filter to distinguish between timeouts and user-requested cancelations.

brandur-stripe commented on src/Stripe.net/Infrastructure/Public/RequestTelemetry.cs:
```diff
@@ -44,8 +50,8 @@ public void MaybeAddTelemetryHeader(HttpHeaders headers)
-        /// <param name="durationMs">The request duration, in milliseconds.</param>
-        public void MaybeEnqueueMetrics(HttpResponseMessage response, long durationMs)
+        /// <param name="duration">The request duration.</param>
+        public void MaybeEnqueueMetrics(HttpResponseMessage response, TimeSpan duration)
```
_Way_ nicer.

brandur-stripe commented on src/Stripe.net/Infrastructure/Public/StripeConfiguration.cs:
```diff
@@ -85,6 +95,12 @@ public static string ApiKey
+        /// Gets or sets the maximum number of times that the library will retry requests that
+        /// appear to have failed due to an intermittent problem.
+        /// </summary>
+        public static int MaxNetworkRetries { get; set; } = 0;
```
Man, they let you set defaults for auto-properties now too?! Love how much boilerplate they've stripped out of this language.

brandur-stripe commented on src/Stripe.net/Infrastructure/Public/SystemNetHttpClient.cs:
```diff
@@ -122,5 +167,62 @@ private static string BuildStripeClientUserAgentString()
+                return true;
+            }
+
+            if (statusCode == HttpStatusCode.Conflict)
```
It might be worth adding 503s here and non-`POST` 500s as recently came into stripe-node. I think that's the direction we're going, and we just haven't changed stripe-go yet.

ob-stripe commented:
I've updated `ShouldRetry` to also retry on 503's and non-POST 5xx's, similar to stripe-node.

I also realized there were a few important bits missing, which I also updated:
- we now set an idempotency key for all POST requests, if one is not already provided by the user (it would have been pretty dangerous to ship automatic retries without this!)
- non-timeout connection errors are reported as `HttpRequestException`s. These are now handled like timeout errors.
- I've also improved error handling to be a bit more resilient with unexpected bodies (I was getting `NullPointerException`s in some of my tests because I was returning `{}` as the body and the error handling code tries to access the `error` key in the JSON hash)

ptal @brandur-stripe

brandur-stripe commented:
> * we now set an idempotency key for all POST requests, if one is not already provided by the user (it would have been pretty dangerous to ship automatic retries without this!)

Phew! Glad you caught that one before shipping, haha.

LGTM!
